File: src/frontend/services/apiClients/lore.ts
// Copyright (C) 2026 AugmentedQuill contributors
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

/** Purpose: Keep project-scoped native lore and raw World Info calls together. */

import { deleteJson, fetchJson, postJson, projectEndpoint, putJson } from './shared';

export type LoreStatus = 'canon' | 'belief' | 'proposal';
export type LoreKind =
  'character' | 'location' | 'organization' | 'object' | 'event' | 'rule' | 'other';
export type SelectiveLogic = 'AND_ANY' | 'NOT_ALL' | 'NOT_ANY' | 'AND_ALL';

export type JsonObject = Record<string, unknown>;

export interface LoreScope {
  book_id?: string | null;
  chapter_id?: string | number | null;
  chapter_start?: number | null;
  chapter_end?: number | null;
  scene_id?: string | number | null;
  scene_start?: number | null;
  scene_end?: number | null;
  viewpoint?: string | null;
  timeline_id?: string | null;
  timeline_position?: number | null;
  timeline_start?: number | null;
  timeline_end?: number | null;
}

export interface LoreActivation {
  enabled: boolean;
  constant: boolean;
  primary_keys: string[];
  secondary_keys: string[];
  selective_logic: SelectiveLogic;
  order: number;
  recursive: boolean;
  prevent_recursion: boolean;
  exclude_recursion: boolean;
  case_sensitive?: boolean | null;
  match_whole_words?: boolean | null;
  [key: string]: unknown;
}

export interface LoreRelation extends JsonObject {
  target_id?: string;
  relation?: string;
}

export interface LoreEntry {
  id: string;
  name: string;
  kind: LoreKind | string;
  status: LoreStatus;
  description: string;
  aliases: string[];
  relations: LoreRelation[];
  sources: unknown[];
  scope: LoreScope;
  activation: LoreActivation;
  belief_actor?: string | null;
  book?: string | null;
  raw_record?: JsonObject | null;
  raw_fields: JsonObject;
  [key: string]: unknown;
}

export interface LoreEntryInput {
  name: string;
  kind: LoreKind | string;
  status: LoreStatus;
  description: string;
  aliases: string[];
  relations: LoreRelation[];
  sources: unknown[];
  scope: LoreScope;
  activation: LoreActivation;
  belief_actor?: string | null;
}

export type LoreEntryUpdate = Partial<LoreEntryInput>;

export interface LoreContextRequest {
  scan_text: string;
  scope: LoreScope;
  budget_tokens?: number | null;
  include_beliefs: boolean;
  include_proposals: boolean;
  recursive: boolean;
  max_recursion_steps: number;
}

export interface LoreDecision {
  entry_id: string;
  included: boolean;
  reason: string;
  matched_primary: string[];
  matched_secondary: string[];
  estimated_tokens: number;
  priority: number;
  recursion_step: number;
}

export interface LoreSelectionResult {
  selected: LoreEntry[];
  decisions: LoreDecision[];
  context_text: string;
  estimated_tokens: number;
  budget_tokens?: number | null;
  budget_warning?: string | null;
  warnings: string[];
  unsupported_options: string[];
}

export interface WorldInfoBookSummary {
  name: string;
  unsupported_options: string[];
}

export type WorldInfoPayload = JsonObject & {
  entries: JsonObject | JsonObject[];
};

export interface LoreApi {
  list: (query?: string) => Promise<LoreEntry[]>;
  create: (payload: LoreEntryInput) => Promise<LoreEntry>;
  get: (entryId: string) => Promise<LoreEntry>;
  update: (entryId: string, payload: LoreEntryUpdate) => Promise<LoreEntry>;
  delete: (entryId: string) => Promise<{ ok: boolean }>;
  select: (payload: LoreContextRequest) => Promise<LoreSelectionResult>;
  listWorldInfo: () => Promise<WorldInfoBookSummary[]>;
  getWorldInfo: (bookName: string) => Promise<WorldInfoPayload>;
  importWorldInfo: (
    bookName: string,
    payload: WorldInfoPayload
  ) => Promise<{
    name: string;
    entries: number;
    unsupported_options: string[];
  }>;
  deleteWorldInfo: (bookName: string) => Promise<{ ok: boolean }>;
}

const encodePathPart = (value: string): string => encodeURIComponent(value);

export const createLoreApi = (projectName: string): LoreApi => ({
  list: (query?: string): Promise<LoreEntry[]> => {
    const params = new URLSearchParams();
    if (query?.trim()) params.set('query', query);
    const queryString = params.toString();
    const path = queryString ? `/lore?${queryString}` : '/lore';
    return fetchJson<LoreEntry[]>(
      projectEndpoint(projectName, path),
      undefined,
      'Failed to load lore'
    );
  },

  create: (payload: LoreEntryInput): Promise<LoreEntry> =>
    postJson<LoreEntry>(
      projectEndpoint(projectName, '/lore'),
      payload,
      'Failed to create lore entry'
    ),

  get: (entryId: string): Promise<LoreEntry> =>
    fetchJson<LoreEntry>(
      projectEndpoint(projectName, `/lore/${encodePathPart(entryId)}`),
      undefined,
      'Failed to load lore entry'
    ),

  update: (entryId: string, payload: LoreEntryUpdate): Promise<LoreEntry> =>
    putJson<LoreEntry>(
      projectEndpoint(projectName, `/lore/${encodePathPart(entryId)}`),
      payload,
      'Failed to update lore entry'
    ),

  delete: (entryId: string): Promise<{ ok: boolean }> =>
    deleteJson<{ ok: boolean }>(
      projectEndpoint(projectName, `/lore/${encodePathPart(entryId)}`),
      'Failed to delete lore entry'
    ),

  select: (payload: LoreContextRequest): Promise<LoreSelectionResult> =>
    postJson<LoreSelectionResult>(
      projectEndpoint(projectName, '/lore/select'),
      payload,
      'Failed to select lore context'
    ),

  listWorldInfo: (): Promise<WorldInfoBookSummary[]> =>
    fetchJson<WorldInfoBookSummary[]>(
      projectEndpoint(projectName, '/lore/world-info'),
      undefined,
      'Failed to list World Info books'
    ),

  getWorldInfo: (bookName: string): Promise<WorldInfoPayload> =>
    fetchJson<WorldInfoPayload>(
      projectEndpoint(projectName, `/lore/world-info/${encodePathPart(bookName)}`),
      undefined,
      'Failed to export World Info book'
    ),

  importWorldInfo: (
    bookName: string,
    payload: WorldInfoPayload
  ): Promise<{
    name: string;
    entries: number;
    unsupported_options: string[];
  }> =>
    postJson<{
      name: string;
      entries: number;
      unsupported_options: string[];
    }>(
      projectEndpoint(projectName, `/lore/world-info/${encodePathPart(bookName)}`),
      payload,
      'Failed to import World Info book'
    ),

  deleteWorldInfo: (bookName: string): Promise<{ ok: boolean }> =>
    deleteJson<{ ok: boolean }>(
      projectEndpoint(projectName, `/lore/world-info/${encodePathPart(bookName)}`),
      'Failed to delete World Info book'
    ),
});

File: src/augmentedquill/api/v1/lore.py
# Copyright (C) 2026 StableLlama
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

"""Project-scoped lore CRUD, deterministic selection, and World Info exchange."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException

from augmentedquill.api.v1.dependencies import ProjectDep
from augmentedquill.services.lore.activation import select_project_lore
from augmentedquill.services.lore.models import (
    LoreContextRequest,
    LoreEntry,
    LoreEntryCreate,
    LoreEntryUpdate,
    LoreSelectionResult,
)
from augmentedquill.services.lore.native import (
    async_create_native_lore,
    async_delete_native_lore,
    async_update_native_lore,
    get_native_lore,
    list_native_lore,
)
from augmentedquill.services.lore.storage import (
    async_delete_world_info,
    async_write_world_info,
    list_world_info_books,
    read_world_info,
    unsupported_world_info_options,
)

router = APIRouter(prefix="/projects/{project_name}", tags=["Lore"])


@router.get("/lore/world-info")
async def api_list_world_info(project_dir: ProjectDep) -> list[dict[str, Any]]:
    """List imported raw World Info books and their unsupported options."""

    books: list[dict[str, Any]] = []
    for name in list_world_info_books(project_dir):
        try:
            payload = read_world_info(project_dir, name) or {}
        except ValueError as exc:
            books.append(
                {
                    "name": name,
                    "unsupported_options": [f"{name}: unreadable World Info ({exc})"],
                }
            )
            continue
        books.append(
            {
                "name": name,
                "unsupported_options": unsupported_world_info_options(name, payload),
            }
        )
    return books


@router.get("/lore/world-info/{book_name:path}")
async def api_get_world_info(book_name: str, project_dir: ProjectDep) -> dict[str, Any]:
    """Export a raw World Info book without normalizing unknown fields."""

    try:
        payload = read_world_info(project_dir, book_name)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    if payload is None:
        raise HTTPException(
            status_code=404, detail=f"World Info book '{book_name}' not found"
        )
    return payload


@router.post("/lore/world-info/{book_name:path}")
async def api_import_world_info(
    book_name: str, payload: dict[str, Any], project_dir: ProjectDep
) -> dict[str, Any]:
    """Import a World Info object, retaining all fields and original IDs."""

    try:
        stored = await async_write_world_info(project_dir, book_name, payload)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return {
        "name": book_name,
        "entries": len(stored.get("entries", {})),
        "unsupported_options": unsupported_world_info_options(book_name, stored),
    }


@router.delete("/lore/world-info/{book_name:path}")
async def api_delete_world_info(
    book_name: str, project_dir: ProjectDep
) -> dict[str, bool]:
    """Delete an imported World Info book."""

    if not await async_delete_world_info(project_dir, book_name):
        raise HTTPException(
            status_code=404, detail=f"World Info book '{book_name}' not found"
        )
    return {"ok": True}


@router.get("/lore", response_model=list[LoreEntry])
async def api_list_lore(
    project_dir: ProjectDep, query: str | None = None
) -> list[LoreEntry]:
    """List native Sourcebook lore and its explicit status/scope metadata."""

    return list_native_lore(project_dir, query=query)


@router.post("/lore", response_model=LoreEntry, status_code=201)
async def api_create_lore(
    payload: LoreEntryCreate, project_dir: ProjectDep
) -> LoreEntry:
    """Create a native Sourcebook entry with lore metadata."""

    try:
        return await async_create_native_lore(project_dir, payload)
    except (ValueError, KeyError) as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("/lore/{entry_id:path}", response_model=LoreEntry)
async def api_get_lore(entry_id: str, project_dir: ProjectDep) -> LoreEntry:
    """Get a native lore entry by its stable ID or current Sourcebook name."""

    entry = get_native_lore(project_dir, entry_id)
    if entry is None:
        raise HTTPException(
            status_code=404, detail=f"Lore entry '{entry_id}' not found"
        )
    return entry


@router.put("/lore/{entry_id:path}", response_model=LoreEntry)
async def api_update_lore(
    entry_id: str, payload: LoreEntryUpdate, project_dir: ProjectDep
) -> LoreEntry:
    """Update Sourcebook content and/or lore metadata without losing its ID."""

    try:
        return await async_update_native_lore(project_dir, entry_id, payload)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.delete("/lore/{entry_id:path}")
async def api_delete_lore(entry_id: str, project_dir: ProjectDep) -> dict[str, bool]:
    """Delete a native lore entry."""

    if not await async_delete_native_lore(project_dir, entry_id):
        raise HTTPException(
            status_code=404, detail=f"Lore entry '{entry_id}' not found"
        )
    return {"ok": True}


@router.post("/lore/select", response_model=LoreSelectionResult)
async def api_select_lore(
    payload: LoreContextRequest, project_dir: ProjectDep
) -> LoreSelectionResult:
    """Return deterministic lore context and an inclusion/exclusion inspector."""

    return select_project_lore(
        project_dir,
        scan_text=payload.scan_text,
        scope=payload.scope,
        budget_tokens=payload.budget_tokens,
        include_beliefs=payload.include_beliefs,
        include_proposals=payload.include_proposals,
        recursive=payload.recursive,
        max_recursion_steps=payload.max_recursion_steps,
    )
